File: Ruby/lib/parrot.rb
class Parrot < Struct.new(:type, :number_of_coconuts, :voltage, :nailed)
  PARROTS = {
    european_parrot: ->(_parrot) { 12 },
    african_parrot: ->(parrot) { 12 - (9 * parrot.number_of_coconuts) },
    norwegian_blue_parrot: ->(parrot) { parrot.voltage * 12 },
  }

  def speed
    return 0 if nailed

    PARROTS[type].call(self).clamp(0..24)
  end
end

File: JavaScript/parrot.js
export const PARROT_TYPES = {
  EUROPEAN: "EUROPEAN",
  AFRICAN: "AFRICAN",
  NORWEGIAN_BLUE: "NORWEGIAN_BLUE",
};

export class Parrot {
  constructor(type, numberOfCoconuts, voltage, isNailed) {
    this.type = type;
    this.numberOfCoconuts = numberOfCoconuts;
    this.voltage = voltage;
    this.isNailed = isNailed;
  }

  getSpeed() {
    if (this.nailed) return 0;

    switch (this.type) {
      case PARROT_TYPES.EUROPEAN:
        return 12;
      case PARROT_TYPES.AFRICAN:
        return Math.max(12 - 9 * this.numberOfCoconuts, 0);
      case PARROT_TYPES.NORWEGIAN_BLUE:
        return Math.min(this.voltage * 12, 24);
    }
  }
}
